# recursion-factorial-process-iterative
> [TABLE OF CONTENTS](https://github.com/E6985/E2746/blob/master/README.md)
***

- any problem that can be solved with - recursion - can also be solved with an - iterative implementation - preferred way to calculate a factorial in this instance as opposed to the - recursive implementation [6]

```python
class str(object='')
class str(object=b'', encoding='utf-8', errors='strict')
```
> [Python Documentation](https://web.archive.org/web/20200503112950/https://docs.python.org/3/library/functions.html#func-str)

```python
str.format(*args, **kwargs)
```
> [Python Documentation](https://web.archive.org/web/20200502135931/https://docs.python.org/3/library/stdtypes.html#string-methods)

In [1]:
nFindFact=5
"factorial: {}".format(nFindFact)

'factorial: 5'

In [2]:
# define function - fFactorial
def fFactorial(nParNonNegInt):
    nResult=1
    while nParNonNegInt>1:
        nResult*=nParNonNegInt
        nParNonNegInt-=1
    return nResult

In [3]:
# call function - fFactorial
fFactorial(nFindFact)

120

- function signature - fFactorial - parameter - nParNonNegInt - five [6]

```python
def fFactorial(nParNonNegInt):
```

- nResult - assigned the value - one [6]

```python
    nResult=1
```

In [4]:
nBase=1
"get to {}".format(nBase)

'get to 1'

- **while-1** - nParNonNegInt - five - greater than one - True [6]

```python
    while nParNonNegInt>1:
```

- nResult - assigned - nParNonNegInt - five - times - nResult - one - 5 [6]

```python
        nResult*=nParNonNegInt
```

- decrement - nParNonNegInt - five - by one - 4 [6]

```python
        nParNonNegInt-=1
```

- return to previous step [6]

In [5]:
"loop running: {}".format(nFindFact>1)

'loop running: True'

In [6]:
"formula: {}.{}".format(nFindFact,nBase)

'formula: 5.1'

In [7]:
nBase*=nFindFact
"running total: {}".format(nBase)

'running total: 5'

In [8]:
nFindFact-=1
"factorial: {}".format(nFindFact)

'factorial: 4'

- **while-2** - nParNonNegInt - four - greater than one - True [6]

```python
    while nParNonNegInt>1:
```

- nResult - assigned - nParNonNegInt - four - times - nResult - tweny - 20 [6]

```python
        nResult*=nParNonNegInt
```

- decrement - nParNonNegInt - four - by one - 3 [6]

```python
        nParNonNegInt-=1
```

- return to previous step [6]

In [9]:
"loop running: {}".format(nFindFact>1)

'loop running: True'

In [10]:
"formula: {}.{}".format(nFindFact,nBase)

'formula: 4.5'

In [11]:
nBase*=nFindFact
"running total: {}".format(nBase)

'running total: 20'

In [12]:
nFindFact-=1
"factorial: {}".format(nFindFact)

'factorial: 3'

- **while-3** - nParNonNegInt - three - greater than one - True [6]

```python
    while nParNonNegInt>1:
```

- nResult - assigned - nParNonNegInt - three - times - nResult - twenty - 60 [6]

```python
        nResult*=nParNonNegInt
```

- decrement - nParNonNegInt - three - by one - 2 [6]

```python
        nParNonNegInt-=1
```

- return to previous step [6]

In [13]:
"loop running: {}".format(nFindFact>1)

'loop running: True'

In [14]:
"formula: {}.{}".format(nFindFact,nBase)

'formula: 3.20'

In [15]:
nBase*=nFindFact
"running total: {}".format(nBase)

'running total: 60'

In [16]:
nFindFact-=1
"factorial: {}".format(nFindFact)

'factorial: 2'

- **while-4** - nParNonNegInt - two - greater than one - True [6]

```python
    while nParNonNegInt>1:
```

- nResult - assigned - nParNonNegInt - two - times - nResult - sixty - 120 [6]

```python
        nResult*=nParNonNegInt
```

- decrement - nParNonNegInt - two - by one - 1 [6]

```python
        nParNonNegInt-=1
```

- return to previous step [6]

In [17]:
"loop running: {}".format(nFindFact>1)

'loop running: True'

In [18]:
"formula: {}.{}".format(nFindFact,nBase)

'formula: 2.60'

In [19]:
nBase*=nFindFact
"running total: {}".format(nBase)

'running total: 120'

- exit **while** loop - nParNonNegInt - one - greater than one - False [6]

```python
    while nParNonNegInt>1:
```

- function - fFactorial - return - nResult - return control to caller [6]

```python
    return nResult
```

***
# END